src/command/dev/watcher.rs

```diff
@@ -55,6 +62,57 @@ impl SubgraphSchemaWatcher {
         )
     }
 
+    pub fn new_from_sdl(
+        subgraph_key: SubgraphKey,
+        sdl: String,
+        message_sender: FollowerMessenger,
+    ) -> RoverResult<Self> {
+        Ok(Self {
+            schema_watcher_kind: SubgraphSchemaWatcherKind::Once(sdl),
+            subgraph_key,
+            message_sender,
+        })
+    }
+
+    pub fn new_from_graph_ref(
+        graph_ref: &str,
+        routing_url: Option<Url>,
+        subgraph_name: String,
+        message_sender: FollowerMessenger,
+        client: &StudioClient,
+    ) -> RoverResult<Self> {
+        // given a graph_ref and subgraph, run subgraph fetch to
+        // obtain SDL and add it to subgraph_definition.
+        let response = fetch::run(
+            SubgraphFetchInput {
+                graph_ref: GraphRef::from_str(graph_ref)?,
+                subgraph_name: subgraph_name.clone(),
+            },
+            client,
+        )
+        .map_err(RoverError::from)?;
+        let routing_url = match (routing_url, response.sdl.r#type) {
+            (Some(routing_url), _) => routing_url,
+            (
+                None,
+                rover_client::shared::SdlType::Subgraph {
+                    routing_url: Some(graph_registry_routing_url),
+                },
+            ) => graph_registry_routing_url.parse().context(format!(
+                "Could not parse graph registry routing url {}",
+                graph_registry_routing_url
+            ))?,
+            (None, _) => {
+                return Err(anyhow!("Could not find routing URL in GraphOS for subgraph {subgraph_name}, try setting `routing_url`").into());
```

Comment:
probably good to create a proper `RoverError` with a `RoverErrorSuggestion` here